File: language-textmate/src/main/java/io/github/rosemoe/sora/langs/textmate/registry/ThemeRegistry.java
/*
 *    sora-editor - the awesome code editor for Android
 *    https://github.com/Rosemoe/sora-editor
 *    Copyright (C) 2020-2024  Rosemoe
 *
 *     This library is free software; you can redistribute it and/or
 *     modify it under the terms of the GNU Lesser General Public
 *     License as published by the Free Software Foundation; either
 *     version 2.1 of the License, or (at your option) any later version.
 *
 *     This library is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public
 *     License along with this library; if not, write to the Free Software
 *     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
 *     USA
 *
 *     Please contact Rosemoe by email 2073412493@qq.com if you need
 *     additional information or have any questions
 */
package io.github.rosemoe.sora.langs.textmate.registry;

import org.eclipse.jdt.annotation.NonNull;
import org.eclipse.jdt.annotation.Nullable;
import org.eclipse.tm4e.core.registry.IThemeSource;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import io.github.rosemoe.sora.langs.textmate.registry.model.ThemeModel;

public class ThemeRegistry {

    private static ThemeRegistry instance;

    private final List<ThemeChangeListener> allListener = new ArrayList<>();

    private final List<ThemeModel> allThemeModel = new ArrayList<>();

    private ThemeModel currentThemeModel;

    public synchronized static ThemeRegistry getInstance() {
        if (instance == null) {
            instance = new ThemeRegistry();
        }
        return instance;
    }

    public ThemeRegistry() {
        currentThemeModel = ThemeModel.EMPTY;
    }


    public void loadTheme(IThemeSource themeSource) throws Exception {
        loadTheme(themeSource, true);
    }

    public void loadTheme(IThemeSource themeSource, boolean isCurrentTheme) throws Exception {
        loadTheme(new ThemeModel(themeSource), isCurrentTheme);
    }

    public void loadTheme(ThemeModel themeModel) throws Exception {
        loadTheme(themeModel, true);
    }

    public synchronized void loadTheme(ThemeModel themeModel, boolean isCurrentTheme) throws Exception {
        if (!themeModel.isLoaded()) {
            themeModel.load();
        }
        var theme = findThemeByThemeName(themeModel.getName());
        if (theme != null) {
            setTheme(theme);
            return;
        }
        allThemeModel.add(themeModel);
        if (isCurrentTheme) {
            setTheme(themeModel);
        }
    }


    @Nullable
    public ThemeModel findThemeByFileName(@NonNull String name) {
        for (var themeModel : allThemeModel) {
            if (themeModel.getName().equals(name)) {
                return themeModel;
            }
        }
        return null;
    }

    @Nullable
    public ThemeModel findThemeByThemeName(@NonNull String name) {
        for (var themeModel : allThemeModel) {
            var rawTheme = themeModel.getRawTheme();

            if (rawTheme == null) {
                continue;
            }

            if (name.equals(rawTheme.getName())) {
                return themeModel;
            }
        }
        return null;
    }

    public synchronized boolean setTheme(@NonNull String name) {
        var targetModel = findThemeByFileName(name);

        if (targetModel != null) {
            setTheme(targetModel);
            return true;
        }

        // need?

        targetModel = findThemeByThemeName(name);

        if (targetModel != null) {
            setTheme(targetModel);
            return true;
        }

        return false;

    }

    public void setTheme(ThemeModel theme) {
        currentThemeModel = theme;

        if (!allThemeModel.contains(theme)) {
            allThemeModel.add(theme);
        }

        if (!theme.isLoaded()) {
            try {
                theme.load();
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }
        dispatchThemeChange(currentThemeModel);
    }


    public ThemeModel getCurrentThemeModel() {
        return currentThemeModel;
    }

    private void dispatchThemeChange(ThemeModel targetThemeModel) {
        for (var listener : allListener) {
            listener.onChangeTheme(targetThemeModel);
        }
    }

    public boolean hasListener(ThemeChangeListener themeChangeListener) {
        return allListener.contains(themeChangeListener);
    }

    public synchronized void addListener(ThemeChangeListener themeChangeListener) {
        allListener.add(themeChangeListener);
    }

    public synchronized void removeListener(ThemeChangeListener themeChangeListener) {
        allListener.remove(themeChangeListener);
    }


    public void dispose() {
        allListener.clear();
    }


    @FunctionalInterface
    public interface ThemeChangeListener {
        void onChangeTheme(ThemeModel newTheme);
    }


}

File: language-textmate/src/main/java/io/github/rosemoe/sora/langs/textmate/registry/ThemeRegistry.kt
/*
 *    sora-editor - the awesome code editor for Android
 *    https://github.com/Rosemoe/sora-editor
 *    Copyright (C) 2020-2024  Rosemoe
 *
 *     This library is free software; you can redistribute it and/or
 *     modify it under the terms of the GNU Lesser General Public
 *     License as published by the Free Software Foundation; either
 *     version 2.1 of the License, or (at your option) any later version.
 *
 *     This library is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public
 *     License along with this library; if not, write to the Free Software
 *     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
 *     USA
 *
 *     Please contact Rosemoe by email 2073412493@qq.com if you need
 *     additional information or have any questions
 */
package io.github.rosemoe.sora.langs.textmate.registry

import io.github.rosemoe.sora.langs.textmate.registry.model.ThemeModel
import org.eclipse.tm4e.core.registry.IThemeSource

class ThemeRegistry {
    private val allListener = arrayListOf<ThemeChangeListener>()

    private val allThemeModel = arrayListOf<ThemeModel>()

    private var _currentThemeModel = ThemeModel.EMPTY
    var currentThemeModel: ThemeModel
        get() = _currentThemeModel
        set(value) {
            _currentThemeModel = value
            if (!allThemeModel.contains(value)) {
                allThemeModel.add(value)
            }
            if (!value.isLoaded) {
                try {
                    value.load()
                } catch (e: Exception) {
                    throw RuntimeException(e)
                }
            }
            dispatchThemeChange(_currentThemeModel)
        }


    @JvmOverloads
    @Throws(Exception::class)
    fun loadTheme(themeSource: IThemeSource, isCurrentTheme: Boolean = true) {
        loadTheme(ThemeModel(themeSource), isCurrentTheme)
    }

    @JvmOverloads
    @Synchronized
    @Throws(Exception::class)
    fun loadTheme(themeModel: ThemeModel, isCurrentTheme: Boolean = true) {
        if (!themeModel.isLoaded) {
            themeModel.load()
        }
        val theme = findThemeByThemeName(themeModel.name)
        if (theme != null) {
            currentThemeModel = theme
            return
        }
        allThemeModel.add(themeModel)
        if (isCurrentTheme) {
            currentThemeModel = themeModel
        }
    }


    fun findThemeByFileName(name: String): ThemeModel? {
        return allThemeModel.find { it.name == name }
    }

    fun findThemeByThemeName(name: String): ThemeModel? {
        return allThemeModel.find { it.rawTheme?.name == name }
    }

    @Synchronized
    fun setTheme(name: String): Boolean {
        val targetModel = findThemeByFileName(name)
            ?: findThemeByThemeName(name)
        return if (targetModel != null) {
            currentThemeModel = targetModel
            true
        } else {
            false
        }
    }

    private fun dispatchThemeChange(targetThemeModel: ThemeModel) {
        allListener.forEach { it.onChangeTheme(targetThemeModel) }
    }

    @Synchronized
    fun addListener(themeChangeListener: ThemeChangeListener) {
        if (allListener.contains(themeChangeListener)) return
        allListener.add(themeChangeListener)
    }

    @Synchronized
    fun removeListener(themeChangeListener: ThemeChangeListener?) {
        allListener.remove(themeChangeListener)
    }


    fun dispose() {
        allListener.clear()
    }


    fun interface ThemeChangeListener {
        fun onChangeTheme(newTheme: ThemeModel)
    }

    companion object {
        private var instance: ThemeRegistry? = null

        @JvmStatic
        fun getInstance(): ThemeRegistry {
            if (instance == null) {
                instance = ThemeRegistry()
            }
            return instance!!
        }
    }
}
